docker_container: do not split command on commas

agshew opened the pull request:
##### SUMMARY
If a command string includes commas, then it is split into a list and converted back to a string without commas. This non-intuitive behavior was caused by defining the type of the command argument as a list. Changing the type to raw and only joining command if it is a list retains current list behavior and keeps commas in strings.

Fixes #24430  

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
docker/docker_container

##### ANSIBLE VERSION
```
ansible 2.4.0 (issue_24430_command_splits_commas 55bace6f9b) last updated 2017/05/22 19:46:19 (GMT +100)
  config file = 
  configured module search path = [u'/home/osboxes/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/osboxes/src/ansible/lib/ansible
  executable location = /home/osboxes/src/ansible/bin/ansible
  python version = 2.7.12+ (default, Sep 17 2016, 12:08:02) [GCC 6.2.0 20160914]
```


##### ADDITIONAL INFORMATION
Tested with the following commands:

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
./hacking/test-module -m lib/ansible/modules/cloud/docker/docker_container.py -a \
"{ \
  \"name\": \"etcd\", \
  \"image\": \"quay.io/coreos/etcd\", \
  \"state\": \"started\", \
  \"volumes\": [ \
    \"/data\" \
  ], \
  \"published_ports\": [ \
    4001, \
    2380, \
    2379 \
  ], \
  \"command\": \"/usr/local/bin/etcd  -name etcd{{ node }} -advertise-client-urls http://{{ myip }}:2379,http://{{ myip }}:4001 -listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 -initial-advertise-peer-urls http://{{ myip }}:2380 -listen-peer-urls http://0.0.0.0:2380 -initial-cluster-token etcd-cluster-1 -initial-cluster etcd0=http://{{ groups['kube-nodes'][0] }}:2380,etcd1=http://{{ groups['kube-nodes'][1] }}:2380,etcd2=http://{{ groups['kube-nodes'][2] }}:2380 -initial-cluster-state new\" \
}"
```

```
./hacking/test-module -m lib/ansible/modules/cloud/docker/docker_container.py -a \
"{ \
  \"name\": \"etcd\", \
  \"image\": \"quay.io/coreos/etcd\", \
  \"state\": \"started\", \
  \"volumes\": [ \
    \"/data\" \
  ], \
  \"published_ports\": [ \
    4001, \
    2380, \
    2379 \
  ], \
  \"command\": [ \
    \"/usr/local/bin/etcd\", \
    \"-name etcd{{ node }}\", \
    \"-advertise-client-urls http://{{ myip }}:2379,http://{{ myip }}:4001\", \
    \"-listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001\", \
    \"-initial-advertise-peer-urls http://{{ myip }}:2380\", \
    \"-listen-peer-urls http://0.0.0.0:2380\", \
    \"-initial-cluster-token etcd-cluster-1\", \
    \"-initial-cluster etcd0=http://{{ groups['kube-nodes'][0] }}:2380,etcd1=http://{{ groups['kube-nodes'][1] }}:2380,etcd2=http://{{ groups['kube-nodes'][2] }}:2380\", \
    \"-initial-cluster-state new\" \
  ], \
}"
```

agshew committed:
docker_container: do not split command on commas

Fix issue #24430

sivel commented:
Wouldn't this break backwards compatibility for those relying on the current behavior?

You can avoid your current problem by doing:

```
command:
  - some, command, with, commas, here
```

agshew committed:
docker_container: document command parameter

Prior behavior was strange, splitting strings on commas
only to join them again ... replacing commas with spaces.

FYI, docker-py accepts strings or lists, using shlex.split
on strings ... splitting on spaces while respecting things
like quotes.

https://github.com/docker/docker-py/blob/master/docker/api/container.py
https://github.com/docker/docker-py/blob/master/docker/types/containers.py
https://github.com/docker/docker-py/blob/master/docker/utils/utils.py

More info regarding Dockerfile syntax for CMD strings and lists
can be found at https://docs.docker.com/engine/reference/builder/#cmd

abadger commented:
We talked about this online and decided that the new syntax is backwards incompatible but it solves a bug and it follows the syntax in an equivalent Dockerfile more closely.  Hopefully the documentation will be sufficient to warn people about the change.
